### PR TITLE
Remove all API deprecated methods/classes/objects

### PR DESCRIPTION
**Pull request by Shanedell**
Remove all API deprecated methods/classes/objects

- daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala: Use withDebugger and withDebugging instead of setDebugger and setDebugging. Rename function setupDebugOrTrace to withDebugOrTrace, also update it to return the processor to be used.
- daffodil-core/src/main/scala/org/apache/daffodil/compiler/Compiler.scala: Remove deprecated functions.
- daffodil-core/src/test/scala/org/apache/daffodil/util/TestUtils.scala: Remove deprecated functions. Also add withDebugger and withDebugging functions.
- daffodil-io/src/main/scala/org/apache/daffodil/io/InputSourceDataInputStream.scala: Remove isAtEnd function.
- daffodil-io/src/test/scala/org/apache/daffodil/io/TestInputSourceDataInputStream8.scala: Remove deprecated functions.
- daffodil-japi/src/main/java/org/apache/daffodil/japi/logger/LogLevel.java: Delete file as nothing in it is used.
- daffodil-japi/src/main/scala/org/apache/daffodil/japi/Daffodil.scala: Remove deprecated functions and unused imports.
- daffodil-japi/src/main/scala/org/apache/daffodil/japi/infoset/Infoset.scala: Remove deprecated functions.
- daffodil-japi/src/main/scala/org/apache/daffodil/japi/logger/Logger.scala: Delete file as nothing in it is used.
- daffodil-japi/src/test/java/org/apache/daffodil/example/TestJavaAPI.java: Rename test to be with instead of set.
- daffodil-lib/src/main/scala/org/apache/daffodil/api/Diagnostic.scala: Remove deprecated function isAtEnd.
- daffodil-propgen/src/main/scala/org/apache/daffodil/propGen/TunableGenerator.scala: Rename setTunable and setTunables to be withTunable and withTunables.
- daffodil-runtime1/src/main/scala/org/apache/daffodil/api/DFDLParserUnparser.scala: Remove deprecated functions. Also add withDebugger and withDebugging functions.
- daffodil-runtime1/src/main/scala/org/apache/daffodil/dsom/CompiledExpression1.scala: Remove deprecated constant.
- daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/JsonInfosetInputter.scala: Remove deprecated this constructor. Also remove Either from default constructor as only InputStream used, remove private so default constructor is callable.
- daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/JsonInfosetOutputter.scala: Remove dummy value and deprecated this constructor.
- daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/XMLTextInfosetInputter.scala: Remove deprecated this constructor. Also remove Either from default constructor as only InputStream used, remove private so default constructor is callable..
- daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/XMLTextInfosetOutputter.scala: Remove dummy value and deprecated this constructor.
- daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataLoc.scala: Remove deprecated isAtEnd function.
- daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataProcessor.scala: Remove deprecated functions and update var to val for some variables.
- daffodil-runtime2/src/main/scala/org/apache/daffodil/runtime2/Runtime2DataProcessor.scala: Remove deprecated functions, coverage on and off comments. Also add withDebugger and withDebugging functions.
- daffodil-sapi/src/main/scala/org/apache/daffodil/sapi/Daffodil.scala: Remove deprecated functions.
- daffodil-sapi/src/main/scala/org/apache/daffodil/sapi/infoset/Infoset.scala: Remove deprecated functions.
- daffodil-sapi/src/main/scala/org/apache/daffodil/sapi/logger/Logger.scala: Delete file as nothing in it is used.
- daffodil-tdml-lib/src/main/scala/org/apache/daffodil/tdml/processor/TDMLDFDLProcessor.scala: Remove deprecated functions.
- daffodil-tdml-processor/src/main/scala/org/apache/daffodil/tdml/processor/DaffodilTDMLDFDLProcessor.scala: Remove deprecated functions.
- daffodil-tdml-processor/src/main/scala/org/apache/daffodil/tdml/processor/Runtime2TDMLDFDLProcessor.scala: Remove deprecated functions.
- daffodil-lib/src/main/scala/org/apache/daffodil/processors/Runtime.scala: Delete file as HasSetDebugger no longer used.
- daffodil-tdml-lib/src/main/scala/org/apache/daffodil/tdml/TDMLRunner.scala: Remove references to HasSetDebugger as it was removed.

[DAFFODIL-2743](https://issues.apache.org/jira/projects/DAFFODIL/issues/DAFFODIL-2743?filter=allissues)

### CODE REVIEW

File: daffodil-japi/src/main/scala/org/apache/daffodil/japi/Daffodil.scala
Comment (stevedlawrence) [inline daffodil-japi/src/main/scala/org/apache/daffodil/japi/Daffodil.scala:850]: This `deprecatedOutput` variable can be removed. This was originally to maintain backwards compatibility with some of the deprecated `parse()` and below `result()` functions, but with those removed this can now be removed. Same in the Scala API.

Comment (Shanedell) [reply daffodil-japi/src/main/scala/org/apache/daffodil/japi/Daffodil.scala:850]: Should this go for any variable that has `deprecated` in its name?

**Comment by stevedlawrence** [comment]
> I also want to know why we didn't see compilation warnings alerting us that these deprecated methods still were being called

I don't think we called any deprecated methods, even in unit tests. Probably not a great idea from a code coverage perspective, but they were deprecated, so maybe it wasn't super important. We current set the `-deprecation` flag in scalacOptions, so I think  deprecation warnings are treated as errors, which is why we don't have any tests calling those functions, or else they would error.

**Comment by tuxji** [comment]
> We current set the `-deprecation` flag in scalacOptions, so I think deprecation warnings are treated as errors, which is why we don't have any tests calling those functions, or else they would error.

OK, I see, the compiler just didn't tell us that some javadocs or scaladocs still referred to deprecated functions.  Treating deprecation warnings as errors is fine with me.  Now the integration tests are failing and will need to be fixed as well.

File: daffodil-japi/src/main/scala/org/apache/daffodil/japi/Daffodil.scala
Comment (Shanedell) [reply daffodil-japi/src/main/scala/org/apache/daffodil/japi/Daffodil.scala:850]: Is `separatorPolicy` also deprecated?

Comment (Shanedell) [reply daffodil-japi/src/main/scala/org/apache/daffodil/japi/Daffodil.scala:850]: Should all the items in `DeprecatedProperty` be removed as well

Comment (stevedlawrence) [reply daffodil-japi/src/main/scala/org/apache/daffodil/japi/Daffodil.scala:850]: `separatorPolicy` is a legacy property that is deprecated. Seems reasonable to remove support for all these deprecated properties as well, unless we want to do that separately. Maybe we only want to remove deprecated APIs for now since their use is likely pretty rare, and remove support for deprecated schema properties at some other point in the future since many schemas may still use these.

Comment (Shanedell) [reply daffodil-japi/src/main/scala/org/apache/daffodil/japi/Daffodil.scala:850]: That would make sense to me because a lot of the stuff I removed already wasn't being used by anything.

Comment (Shanedell) [reply daffodil-japi/src/main/scala/org/apache/daffodil/japi/Daffodil.scala:850]: @stevedlawrence Do you think anything other than `deprecatedOutput` and changing the default value of `allowExpressionResultCoercion` to false or do you think it is mostly schema stuff?

Comment (stevedlawrence) [reply daffodil-japi/src/main/scala/org/apache/daffodil/japi/Daffodil.scala]: I think it depends what our goal here is. If the goal is to ensure we don't break existing schemas and we're only removing deprecated APIs (which could require code changes but no schema changes), then we shouldn't change this tunable--changing it may require schema changes if schemas rely on this coercion.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataProcessor.scala
Comment (stevedlawrence) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataProcessor.scala:148]: There's a comment just above that says:
```scala
  //
  // Once the deprecated API goes away, these var can become val.
  //
```
We should probably make that change now.

**Comment by Shanedell** [comment]
> Your pull request description needs a body as well as the title and the DAFFODIL-2743 line. The body should say which Scala classes had to be changed because they still had references to deprecated methods. I also want to know why we didn't see compilation warnings alerting us that these deprecated methods still were being called. I think it would be in scope for this PR to (temporarily or permanently) add a compilation option to enable deprecation warnings so we can see what other problems are there.

@tuxji Do you want the specific class names and which methods were updated, just the class names that need updated/removed or just the file name?

File: daffodil-japi/src/main/scala/org/apache/daffodil/japi/Daffodil.scala
Comment (Shanedell) [reply daffodil-japi/src/main/scala/org/apache/daffodil/japi/Daffodil.scala]: That makes sense to me. I think it would be best to do the schema changes in another issue/PR as I feel this one is quite large enough that splitting out the other changes would make sense.

**Comment by tuxji** [comment]
> Do you want the specific class names and which methods were updated, just the class names that need updated/removed or just the file name?

Shane, the best level of detail to put in a pull request is somewhat subjective.  I personally review my own pull requests one file at a time and write a brief message about each file because doing so forces me to think about why the PR changes each file and sometimes reminds me of something I may have overlooked.  Here is an example I see in your own pull request.  The first changed file is Main.scala so I would write something like...

`Main.scala: Call DataProcessor methods instead of deprecated HasSetDebugger methods.`

I see two issues immediately when I look at Main.scala.  The first issue is that when I search all the files changed for other mentions of HasSetDebugger, I don't see HasSetDebugger's definition, only that it is no longer one of DataProcessor's traits.  I know immediately that the PR has not removed HasSetDebugger's definition yet (FYI, you can find HasSetDebugger's definition in the file daffodil-lib/src/main/scala/org/apache/daffodil/processors/Runtime.scala).  The second issue I see is that you can't simply replace `setDebugg*` calls with `withDebugg*` calls.  The `setDebugg*` calls are mutative - they change the called object's state.  The `withDebugg*` calls are immutative - they return a copy of the called object with the new value instead of the old value, but the called object's state hasn't changed in any way.  This could be why your integration tests are failing.

This example shows how the very act of writing down why your PR changes each file can help you notice problems yourself.

File: daffodil-japi/src/main/scala/org/apache/daffodil/japi/Daffodil.scala
Comment (mbeckerle) [reply daffodil-japi/src/main/scala/org/apache/daffodil/japi/Daffodil.scala]: Or at least a separate change set, so they're more easily spotted and the change backed out if need be.

**Comment by Shanedell** [comment]
@tuxji Your comment makes a lot of sense to me thank you. Also with the Integration tests, is there a way I can run a single test or suite tests at a time instead of all them at once? This would just be to be to test fixes for them timing out.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataProcessor.scala
Comment (Shanedell) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataProcessor.scala:148]: This will be added in the next commit

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/XMLTextInfosetOutputter.scala
Comment (Shanedell) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/XMLTextInfosetOutputter.scala:35]: Should this go for each file it is in? I am seeing this in these files

```txt
daffodil-lib/src/main/scala/org/apache/daffodil/exceptions/Assert.scala --> line 32
daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/JsonInfosetOutputter.scala --> line 28
daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/XMLTextInfosetOutputter.scala --> line 35 (item commented on)
```

**Comment by mbeckerle** [comment]
I put the instructions for running just one test on a wiki page, since I had always forgotten it by the time I needed it again.

https://cwiki.apache.org/confluence/display/DAFFODIL/sbt+Tips+and+Tricks

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/XMLTextInfosetOutputter.scala
Comment (Shanedell) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/XMLTextInfosetOutputter.scala]: I removed this from both the XML and JSON Infoset outputter's as it seemed to not cause issues but did have other effects inside of `Assert.scala`

**Comment by Shanedell** [comment]
@stevedlawrence @mbeckerle @tuxji This should be good for another good look over as I fixed the issue that was causing issues with integration tests.

File: daffodil-propgen/src/main/scala/org/apache/daffodil/propGen/TunableGenerator.scala
Comment (stevedlawrence) [inline daffodil-propgen/src/main/scala/org/apache/daffodil/propGen/TunableGenerator.scala:98]: This functions aren't technically deprecated, but this rename is semi-related to other deprecated functions and is a good change--the generated `setTunable` functions are immutable and so should use the same `with*` naming convention :+1:

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/externalvars/ExternalVariablesLoader.scala
Comment (stevedlawrence) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/externalvars/ExternalVariablesLoader.scala:50]: Does `withExternalVariable` mutate the vmap, or does it create a copy? I assume it creates a copy with the new values, so I would expect this to be something like:
```scala
val newVMap = VariableUtils.withExernalVariables(...)
newVMap
```

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/XMLTextInfosetInputter.scala
Comment (stevedlawrence) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/XMLTextInfosetInputter.scala:191]: Can this remove the `Either`? Seems like we only need the `InputStream` now?

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/VariableMap1.scala
Comment (stevedlawrence) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/VariableMap1.scala:144]: This calls `setExternalVariable` which looks like it mutates the variable map. Because of that this should probably stay as `setExternalVariables` to make it clear this mutates state.

This also related to another comment I made about what calls this function, that comment probably no longer applies.

Note I wonder if there is a bug...do our API `withExternalVariables` functions in the DataProcessor end up calling this `setExternalVariables` function and mutate state instead of copying? We may need to investigate that, but that can be done separately.

File: daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala
Comment (tuxji) [reply daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala:119]: I would go further and remove the source file defining HasSetDebugger:

daffodil-lib/src/main/scala/org/apache/daffodil/processors/Runtime.scala

Once nothing uses a piece of code anymore, it becomes dead code and should be removed.

File: daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala
Comment (tuxji) [reply daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala:633]: I would have suggested the same change too, thanks, Steve.

File: daffodil-japi/src/test/java/org/apache/daffodil/example/TestJavaAPI.java
Comment (tuxji) [inline daffodil-japi/src/test/java/org/apache/daffodil/example/TestJavaAPI.java:974]: Good catch.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/externalvars/ExternalVariablesLoader.scala
Comment (tuxji) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/externalvars/ExternalVariablesLoader.scala:50]: As Steve noticed later on, `VariableUtils.setExternalVariables` was renamed to `VariableUtils.withExternalVariables` but it still mutates the variable map so this renaming change should be reverted.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/VariableMap1.scala
Comment (tuxji) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/VariableMap1.scala:144]: I checked daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataProcessor.scala lines 297, 308, and 332 in the main branch.  They all have similar bodies:

```scala
     val newBindings = loadExternalVariables(extVars)
     copy(externalVars = newBindings)
```

`loadExternalVariables` is defined in the same file and it returns a new data structure, so the `withExternalVariables` methods do not mutate state instead of copying.  There is no need for concern.

**Review by tuxji — COMMENTED** [review]
More changes needed, but we are getting closer to the end.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/XMLTextInfosetInputter.scala
Comment (Shanedell) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/XMLTextInfosetInputter.scala:191]: Removing the `Either` seems to cause issue with the `this` constructor. I can try to get that worked out for both JSON and XML infoset inputters if desired.

File: daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala
Comment (Shanedell) [reply daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala:119]: @tuxji I did not delete `HasSetDebugger` as it still is used by `daffodil-tdml-lib/src/main/scala/org/apache/daffodil/tdml/TDMLRunner.scala`. The `setDebugger` and `setDebugging` methods were set as deprecated here that is why they weren't delete but this can be done if it is something we want to do.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/XMLTextInfosetInputter.scala
Comment (Shanedell) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/XMLTextInfosetInputter.scala:191]: @stevedlawrence To get this to work the `private` keyword need deleted for example for `JsonInfosetInputter` I needed to make the class def be this
```scala
class JsonInfosetInputter(input: java.io.InputStream)
```
instead of having
```scala
class JsonInfosetInputter private (input: java.io.InputStream)
```
would this be how we wanted to do it?

File: daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala
Comment (tuxji) [reply daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala:119]: I looked at TDMLRunner.scala and I don't think it has any good reason why DFDLTestSuite needs to extend HasSetDebugger.  If you remove the import, the extends HasSetDebugger, and the override keyword from the `setDebugger` and `setDebugging` methods and you still can build the code and run the tests without errors, we'll know for sure.

Comment (Shanedell) [reply daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala:119]: Makes sense, doing this worked fine for the tests.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/XMLTextInfosetInputter.scala
Comment (stevedlawrence) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/XMLTextInfosetInputter.scala:191]: Makes sense, we now only need the one constructor so it can be made public.

File: daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala
Comment (tuxji) [inline daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala:933]: In this scope, processor is a mutable var so you may not need `debugProcessor`.  Perhaps line 933 can replace processor exactly like line 931 already replaces processor:

`processor = setupDebugOrTrace(processor.asInstanceOf[DataProcessor], conf)`

That change should work unless the compiler complains about types, and it also makes me think the method name might look better if it was renamed to `withDebugOrTrace`.

There's a region further down where processor is a val, so debugProcessor is really needed there.  Mike, Steve, would you prefer that both regions use a mutable var processor or an immutable val processor and a val debugProcessor?  I think it's best practice to use val and careful variable naming to make it clear a value is changing, but var is probably fine in a small lexical scope especially when processor is replaced 2 times, not just 1 time.

**Review by tuxji — APPROVED** [review]
+1

Changes look good now, although I noticed one region of Main.scala uses a mutable var processor.  One could take advantage of that mutability or leave it alone; I'll let others weigh in on what should be best practice.

File: daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala
Comment (stevedlawrence) [reply daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala:933]: Yeah, the var approach seems reasonable to me in this case. Especially since `debugProcessor` might not even have debugging enabled depending on the provide conf options, so it could potentially be confusing. It's more like `maybeDebugProcessor` which is ugly and still confusing.

An alternative, though maybe a bit more invasive, might be to do what we do with the performance command, which looks like this:

```scala
        val processor = {
          if (performanceOpts.parser.isDefined) {
            createProcessorFromParser(performanceOpts.parser(), performanceOpts.path.toOption, validate)
          } else {
            val tunables = DaffodilTunables.configPlusMoreTunablesMap(performanceOpts.tunables, optDafConfig)
            createProcessorFromSchema(performanceOpts.schema(), performanceOpts.rootNS.toOption, performanceOpts.path.toOption, tunables, validate)
          }
        }.map{ _.withExternalVariables(combineExternalVariables(performanceOpts.vars, optDafConfig)) }
         .map{ _.withValidationMode(validate) }
```
We could just add a `.map { withDebugOrTrace(_, conf) }` to the end of that.

This way all processor setup happens in one spot for each subcommand, the processor variable can always be a val, and is consistent among the different subcommands.

Comment (Shanedell) [reply daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala:933]: I will add the `.map { withDebugOrTrace(_, conf) }`. But with that should I also remove the line that does `var processor = proc` and just use `proc` instead?

Comment (stevedlawrence) [reply daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala:933]: Makes sense, that would use the same patern as the unparse and performance subcommands

Comment (Shanedell) [reply daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala:933]: So `processor` would now need to be `Option[DataProcessor]` instead of `Option[DFDL.DataProcessor]` right?

Comment (Shanedell) [reply daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala:933]: So I was able to get it work like this:

```scala
val processor: Option[DataProcessor] = {
  if (parseOpts.parser.isDefined) {
    createProcessorFromParser(parseOpts.parser(), parseOpts.path.toOption, validate)
  } else {
     val tunables = DaffodilTunables.configPlusMoreTunablesMap(parseOpts.tunables, optDafConfig)
      createProcessorFromSchema(parseOpts.schema(), parseOpts.rootNS.toOption, parseOpts.path.toOption, tunables, validate)
  }
}.map{ _.withExternalVariables(combineExternalVariables(parseOpts.vars, optDafConfig)) }
 .map{ _.withValidationMode(validate) }
 .map{ proc => withDebugOrTrace(proc.asInstanceOf[DataProcessor], conf) }
```

Doing `.map{ withDebugOrTrace(_.asInstanceOf[DataProcessor], conf) }` or `.map{ withDebugOrTrace(_, conf) }` caused errors. Is this what would be desired or would you want me to add `withDebugger` defintions to `DataProcessorBase` trait that you just mentioned?

Comment (stevedlawrence) [reply daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala:933]: I have no strong preference. It might be cleaner if the withDebugger defs were added to the trait though. I'm not sure if there was a reason for those to not be added originally. Might have just been an oversight.

Comment (Shanedell) [reply daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala:933]: So when I add them in I still need to do the `.asInstanceOf[DataProcessor]`, is there something I was missing because up until `withDebugOrTrace` I believe that processor is treated as a `DFDL.DataProcessor` right? So that would always be needed unless converting other methods to switch the type of `DataProcessor` being set right?

Comment (stevedlawrence) [reply daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala:933]: You might need to change `withDebugOrTrace` to accept a `DFDL.DataProcessor` instead of a `DataProcessor`.

Comment (Shanedell) [reply daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala:933]: Oh okay I think I got will push that up once I get it fully tested locally to make sure it works fully.

Comment (Shanedell) [reply daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala]: @tuxji @stevedlawrence The rename of `setupDebugOrTrace` to `withDebugOrTrace`, adding `withDebugger` method to`DataProcessorBase` and updating the way the processor is set is up if you want to look it over.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataProcessor.scala
Comment (stevedlawrence) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataProcessor.scala:135]: Can `validationMode` and `externalVars` become  a val too like above? We don't have the set* functions anymore so nothing should mutate this I think?

**Review by stevedlawrence — APPROVED** [review]
+1 :+1: 

Just one comment about some more parameters that I *think* can become vals.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataProcessor.scala
Comment (stevedlawrence) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataProcessor.scala:135]: Right above this we do `val dpToSave = dp.copy()`. We could change that to something like his:
```scala
val dpToSave = this.copy(
  externalVars = Queue.empty[Binding],
  validationMode = ValidationMode.Off
)
```
That's a fairly small change and I think makes our DataProcessor fully immutable.

Comment (Shanedell) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataProcessor.scala:135]: That makes sense I will throw that in

Comment (Shanedell) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataProcessor.scala]: @stevedlawrence That look good? So since I have two approvals, I should wait till the CI passed, then squash my commits, push up the squashed commit and then merge once CI passes?

Comment (tuxji) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataProcessor.scala]: Yes, all the latest changes look good (thanks, Shane and Steve, for adding them).  Yes, CI has passed so you can squash your commits, push a single squashed commit, and merge (if you have write permission).

Comment (tuxji) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataProcessor.scala]: Yes, the commit message can reuse your nicely written PR description (updated for the latest changes).  That would be very nice for git history.

**Pull request merged**
